### ISSUE TITLE
Child class overrides lambda in parent class

### ISSUE THREAD

**Issue by calvarez-ov** [orfjackal/retrolambda#109]
This was initially reported in the [gradle-retrolambda](https://github.com/evant/gradle-retrolambda/issues/207) project, but the problem appears to come from this project. 

Copying the report here:

---

``` java
public abstract class ParentActivity extends AppCompatActivity {

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(getContentView());
        findViewById(R.id.button_parent).setOnClickListener(view -> Toast.makeText(this, "Parent onClick", Toast.LENGTH_LONG).show());
    }

    public abstract int getContentView();
}
```

``` java
public class MainActivity extends ParentActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        findViewById(R.id.button_child).setOnClickListener(view -> Toast.makeText(this, "Child onClick", Toast.LENGTH_LONG).show());
    }

    @Override
    public int getContentView() {
        return R.layout.activity_main;
    }
}
```

Clicking on button_parent shows a Toast with "Child onClick"

test app: https://github.com/cpalasanu/RetrolambdaTest

---

Using the [cfr decompiler](http://www.benf.org/other/cfr/), we see that the child activity class has a method that overrides the parent activity class:  

**Parent**:

``` java
/*
 * Decompiled with CFR 0_118.
 * 
 * Could not load the following classes:
 *  android.content.Context
 *  android.os.Bundle
 *  android.support.annotation.Nullable
 *  android.support.v7.app.AppCompatActivity
 *  android.view.View
 *  android.view.View$OnClickListener
 *  android.widget.Toast
 */
package com.orange.retrolambdatest;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Toast;
import com.orange.retrolambdatest.ParentActivity$$Lambda$1;

public abstract class ParentActivity
extends AppCompatActivity {
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.setContentView(this.getContentView());
        this.findViewById(2131427412).setOnClickListener(ParentActivity$$Lambda$1.lambdaFactory$(this));
    }

    public abstract int getContentView();

    /* synthetic */ void lambda$onCreate$0(View view) {
        Toast.makeText((Context)this, (CharSequence)"Parent onClick", (int)1).show();
    }
}
```

**Child**:

``` java
/*
 * Decompiled with CFR 0_118.
 * 
 * Could not load the following classes:
 *  android.content.Context
 *  android.os.Bundle
 *  android.view.View
 *  android.view.View$OnClickListener
 *  android.widget.Toast
 */
package com.orange.retrolambdatest;

import android.content.Context;
import android.os.Bundle;
import android.view.View;
import android.widget.Toast;
import com.orange.retrolambdatest.MainActivity$$Lambda$1;
import com.orange.retrolambdatest.ParentActivity;

public class MainActivity
extends ParentActivity {
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.findViewById(2131427413).setOnClickListener(MainActivity$$Lambda$1.lambdaFactory$(this));
    }

    @Override
    public int getContentView() {
        return 2130968602;
    }

    @Override
    /* synthetic */ void lambda$onCreate$0(View view) {
        Toast.makeText((Context)this, (CharSequence)"Child onClick", (int)1).show();
    }
}
```

Both parent and child have the same method `lambda$onCreate$0(View view)`

---

I did some tests building the retrolambda project locally.  It appears the bug appears in this commit: https://github.com/orfjackal/retrolambda/commit/bf932457e45817cdee07f34c8c2059db44848e03 (between releases 2.2.0 and 2.3.0)  The PR for that commit: https://github.com/orfjackal/retrolambda/pull/86

I have not done any investigation to understand what part of that commit introduces the bug.

**Comment by calvarez-ov** [comment]
Before the given commit:
Parent:
```java
    private /* synthetic */ void lambda$onCreate$0(View view) {
        Toast.makeText((Context)this, (CharSequence)"Parent onClick", (int)1).show();
    }

    static /* synthetic */ void access$lambda$0(ParentActivity parentActivity, View view) {
        parentActivity.lambda$onCreate$0(view);
    }
```
Child:
```java
    private /* synthetic */ void lambda$onCreate$0(View view) {
        Toast.makeText((Context)this, (CharSequence)"Child onClick", (int)1).show();
    }

    static /* synthetic */ void access$lambda$0(MainActivity mainActivity, View view) {
        mainActivity.lambda$onCreate$0(view);
    }
```
---

After:
Parent:
```java
    /* synthetic */ void lambda$onCreate$0(View view) {
        Toast.makeText((Context)this, (CharSequence)"Parent onClick", (int)1).show();
    }
```
Child:
```java
    @Override
    /* synthetic */ void lambda$onCreate$0(View view) {
        Toast.makeText((Context)this, (CharSequence)"Child onClick", (int)1).show();
    }
```

**Comment by calvarez-ov** [comment]
Related issue: https://github.com/orfjackal/retrolambda/issues/81

**Comment by calvarez-ov** [comment]
Possibly related (or even a duplicate of) https://github.com/orfjackal/retrolambda/issues/95

**Comment by orfjackal** [comment]
Thanks for your work on fixing this and sorry about taking so long to respond.

The origin of the bug is the `access &= ~ACC_PRIVATE;` line in BackportLambdaInvocations.java which makes private instance methods package-private, thus making them potentially overridable in child classes. I'd like to fix the bug once and for all by removing that line.

One solution is to convert the private instance method into a package-private static method, which won't be overridable by child classes. This can be done safely for lambda implementation methods which only the lambda class will call - for other methods it would fall back to generating a bridge access method.

**Comment by orfjackal** [comment]
I've made a fix for this which converts the private instance methods into package-private static methods. Please try it out. I'll release a new version soon.

**Comment by orfjackal** [comment]
The fix to this has been released in Retrolambda 2.5.0
